### PR TITLE
Fix pv[c]Ensurable caching bug

### PR DESCRIPTION
**Pull request by 2uasimojo**
We were neglecting to clear the pvEnsurable and pvcEnsurable caches when
deleting their respective objects. This resulted in behavior where
creating a SharedVolume, deleting it, then recreating one with the same
name but a different spec would result in the *old* spec being used.

Fix this by explicitly clearing the cache entry if we attempt to delete
the resource (even if that deletion fails).

On the test side, completely rework the test case to be fake- rather
than mock-driven.

(The same principle should be applied throughout so we can hopefully get
rid of the mocked client entirely. That'll be the subject of (a)
different PR(s), tracked by [OSD-4098](https://issues.redhat.com/browse/OSD-4098).)

Jira: None

### CODE REVIEW

**Comment by 2uasimojo** [comment]
Tested this live as well. Works.

File: pkg/controller/sharedvolume/sharedvolume_controller.go [hunk 1]
```diff
@@ -217,19 +217,25 @@ func (r *ReconcileSharedVolume) handleDelete(logger logr.Logger, sharedVolume *a
...
+	// Note that Delete only cares about the NamespacedName of each Ensurable. This matters
+	// because it's theoretically possible that the guts of the PV and/or PVC are out of sync
+	// with what's in the SharedVolume. We specifically don't care if that's true because it's
+	// all going away.
+	// TODO(efried): Move cache cleaning logic into pv[c]_ensurable.go... somehow.
+	// Order matters here. Delete the PVC first...
+	e := pvcEnsurable(sharedVolume)
+	k := svKey(sharedVolume)
+	defer delete(pvcBySharedVolume, k)
+	if err := e.Delete(logger, r.client); err != nil {
```
Comment (jharrington22) [inline pkg/controller/sharedvolume/sharedvolume_controller.go:229]: You will delete the key from the map even if this calls fails, is that what you want?

Comment (2uasimojo) [reply pkg/controller/sharedvolume/sharedvolume_controller.go:229]: Yes it is. I did ponder this while writing it and felt this was the best approach.
1. It's possible the delete could have still occurred on the server.
2. I'm in a delete path, so even though there's a failure, we're clearly *trying* to delete. The retry will recreate the cache entry temporarily until we get here again. And so on.
3. Worst case, we didn't really want to delete (not sure how that would happen) and we'll just recreate the cache entry next time around. The worst thing that happens in that (unlikely) scenario is we do an extra reconcile to bring the cached version up to scratch with what's on the server. Which is fine.

Comment (jharrington22) [reply pkg/controller/sharedvolume/sharedvolume_controller.go:229]: :+1:

**Pull request merged**